### PR TITLE
Add beancount-render

### LINKED ISSUE #20: Generate output
What is required to generate beancount ouput? Should the provided types just implement `Display`?

### PR DESCRIPTION
**Pull request by ThomasdenH**
This adds beancount-render as a separate crate to this repository. I have added some tests for correctness, but some are still excluded for various reasons. Adding more tests in general will probably be a good next focus.

Closes #20.

### CODE REVIEW

**Comment by ThomasdenH** [comment]
I'm not sure whether `std::io::Write` or `std::fmt::Write` is better to implement. Maybe both?

File: beancount-render/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,413 @@
...
+}
+
+pub fn render<W: Write>(w: &mut W, ledger: &Ledger<'_>) -> Result<(), BasicRendererError> {
+    BasicRenderer::default().render(ledger, w)
+}
+
+#[derive(Error, Debug)]
+pub enum BasicRendererError {
+    #[error("an io error occurred")]
+    Fmt(#[from] fmt::Error),
```
Comment (twilco) [inline beancount-render/src/lib.rs:27]: The error message here seems a little inconsistent with variant name (`Fmt`).  Maybe this message could be something like "error formatting message into stream" to more closely match the origin error.  Thoughts?

https://doc.rust-lang.org/std/fmt/struct.Error.html

File: beancount-render/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,15 @@
+[package]
+name = "beancount-render"
+version = "0.1.0"
+authors = ["Thomas <denhollander.thomas@gmail.com>"]
+edition = "2018"
+
+[lib]
```
Comment (twilco) [inline beancount-render/Cargo.toml:7]: Does the empty `[lib]` directive need to be present?

**Comment by twilco** [comment]
> I'm not sure whether std::io::Write or std::fmt::Write is better to implement. Maybe both?

Checking the [std::fmt::Write docs](https://doc.rust-lang.org/std/fmt/trait.Write.html), they say:

> This trait is the type which this modules requires when formatting information. This is similar to the standard library's io::Write trait, but it is only intended for use in libcore.
>
> This trait should generally not be implemented by consumers of the standard library. The write! macro accepts an instance of io::Write, and the io::Write trait is favored over implementing this trait.

So `std::io::Write` seems like a better choice 🙂

File: beancount-render/src/lib.rs (hunk 1 above)
Comment (ThomasdenH) [reply beancount-render/src/lib.rs:27]: Ah, that's an artifact of the `Write` change!

**Comment by ThomasdenH** [comment]
> This approach makes sense if we see ourselves creating multiple different types of renderers, presumably for different types of configurations. Is that what you were thinking?

Yes, there could be different ways of handling whitespace, but that's not very interesting to me. My main intention was indeed so that it could hold a state, to eventually facilitate alignment of currencies. I like the config method as well though. It should then be possible to generate a default config automatically such that the alignment works out for the entire ledger.

**Comment by twilco** [comment]
> My main intention was indeed so that it could hold a state, to eventually facilitate alignment of currencies.

Ah, I see.  Yes, my suggested approach is obviously stateless in its current form.  This might be a dumb question, but it is a necessary one, as I haven't actually _used_ Beancount much -- would state ever be necessary to render Beancount structures?  Or could all cases be simply handled by config points?  I would hope the latter, and your current implementation has no state, so that leads me to believe stateless-ness is possible.  If so, I think that would be my preferred approach, but would love to hear your thoughts.

> It should then be possible to generate a default config automatically such that the alignment works out for the entire ledger.

Yep, exactly!

**Comment by mbudde** [comment]
> > My main intention was indeed so that it could hold a state, to eventually facilitate alignment of currencies.
> 
> Ah, I see. Yes, my suggested approach is obviously stateless in its current form. This might be a dumb question, but it is a necessary one, as I haven't actually _used_ Beancount much -- would state ever be necessary to render Beancount structures? Or could all cases be simply handled by config points? I would hope the latter, and your current implementation has no state, so that leads me to believe stateless-ness is possible. If so, I think that would be my preferred approach, but would love to hear your thoughts.

You need to track root account names to render [this ledger](https://github.com/beancount/beancount/blob/b6cc9d74725c2c460519360eb14c52aa335f1095/beancount/parser/options_test.py#L63-L73) correctly:
```
2014-01-01 open Assets:CA:RBC:Checking

option "name_assets" "Actif"

2014-01-04 open Actif:CA:RBC:CompteChèques
```

**Comment by twilco** [comment]
Hmm, OK.  Since state seems to be required, we could do something like this, a modification of what I have above:

```rust
pub trait Render<W: Write> {
    fn render(&self, destination: &mut W, state: &mut RenderState, config: &RenderConfig) -> Result<(), RenderError>;
}

pub struct RenderState {
    // Could maybe be &str, but don't want to mess with lifetimes in this pseudocode
    root_account_name: Option<String>,
    ...more state...
}
```

And each BC construct would be responsible for populating state related to itself as it is rendered, e.g. when the root account name renders itself it sets `root_account_name`.  For consumers of state, they can decide what to do if some state they require is `None`.

Or we could stick with @ThomasdenH's current implementation and track state on the renderer.  Would configuration be stored in the renderer then, too?

What are your thoughts?

**Comment by ThomasdenH** [comment]
I wonder whether all settings are configurations made in the code? For example, the root account name can be considered a render state, but just as well as just a change in the configuration. In which case it might be enough to pass `&mut RenderConfig` to the render function.

I also wonder what to do with the state/configuration afterwards. Would it be good to pass it by value to prevent reuse? It might be unexpected that subsequent renders with the same state/config could yield different results.

As an aside, I have noticed that I seldom want to render a full valid ledger. I might want to generate beancount from a `.csv`, but then I won't open the used accounts. That makes a `render(ledger: &Ledger, ...)` function not as useful as it might sound.

**Comment by twilco** [comment]
> For example, the root account name can be considered a render state, but just as well as just a change in the configuration.

Yeah, the line between "state" and "config" is fuzzy.

> I also wonder what to do with the state/configuration afterwards. Would it be good to pass it by value to prevent reuse?

Good point — agree, assuming the thing we pass into the `render` function is mutable.  Passing something immutable (like a `&RenderConfig`) should still be fine.

> That makes a render(ledger: &Ledger, ...) function not as useful as it might sound.

Yeah, also agree here.  I think such a function would be unnecessary.

---

There has been a lot of good discussion here.  Unless you're re-thinking anything @ThomasdenH, I am becoming inclined to accept this PR as-is.  This provides useful functionality as it stands, and we can evolve the code as we learn more.  Thoughts?

**Comment by ThomasdenH** [comment]
I'll update it to use `io::Write`, then it's probably OK.

**Comment by ThomasdenH** [comment]
I have updated the usage. It's not ideal to use still. Maybe it would be better to have a wrapper type that implements `Display` instead? Something like `Renderable<Directive<'_>>`?

That being said, this PR could probably be merged for now.

**Comment by twilco** [comment]
Yeah, I say let's merge this for now and we can circle back with improvements later.  Thanks again!

**Pull request merged**